contrib/babelfishpg_tsql/src/pltsql_coerce.c

```diff
@@ -888,6 +920,227 @@ run_tsql_best_match_heuristics(int nargs, Oid *input_typeids, FuncCandidateList
+
+Datum
+get_immediate_base_type_of_UDT(PG_FUNCTION_ARGS)
+{
+	Oid			base_type;
+	
+	base_type = get_immediate_base_type_of_UDT_internal(PG_GETARG_OID(0));
+	if (!OidIsValid(base_type))
+		PG_RETURN_NULL();
+
+	PG_RETURN_OID(base_type);
+}
+
+/*
+ * For a given function details, validate whether it is in special function list
+ * and also validate the input argument data types.
+ */
+bool
+validate_special_function(char *func_nsname, char *func_name, int nargs, Oid *input_typeids)
+{
+	tsql_special_function_t    *special_func;
+	bool                        type_match;
+	Oid                         inputTypId, validTypId, baseTypeId;
+	Oid                         sys_varcharoid;
+
+	/* Sanity checks */
+	if (func_name == NULL || (nargs != 0 && input_typeids == NULL))
+		return false;
+
+	/* 
+	 * Special function handling is only for some specific system functions.
+	 * If func_nsname is NULL, consider it to be a "sys".
+	 */
+	if (func_nsname != NULL &&
+		(strlen(func_nsname) != 3 || strncmp(func_nsname, "sys", 3) != 0))
+		return false;
+
+	/* Get Special function details */
+	special_func = NULL;
+	for (int i = 0; i < TOTAL_TSQL_SPECIAL_FUNCTION_COUNT; i++)
+	{
+		if (strcmp(func_name, tsql_special_function_list[i].funcname) == 0
+			&& nargs == tsql_special_function_list[i].nargs)
+		{
+			special_func = &tsql_special_function_list[i];
+			break;
+		}
+	}
+
+	/* If function is not a special function no additional handling required */
+	if (special_func == NULL)
+		return false;
+
+	sys_varcharoid = get_sys_varcharoid();
+	/* Report error in case of invalid argument datatype */
+	for (int i = 0; i < special_func->nargs; i++)
+	{
+		/* for unknown literals consider its type as sys.VARCHAR */
+		inputTypId = (input_typeids[i] == UNKNOWNOID) ? sys_varcharoid : input_typeids[i];
+
+		/* for UDT use its base type for input argument datatype validation */
+		baseTypeId = get_immediate_base_type_of_UDT_internal(inputTypId);
+		if(OidIsValid(baseTypeId))
+		{
+			inputTypId = baseTypeId;
+		}
+
+		type_match = false;
+		for (int j = 0; j < special_func->valid_arg_types[i].len; j++)
+		{
+			if (!OidIsValid(special_func->valid_arg_types[i].valid_types_oid[j]))
+			{
+				/* if common_utility_plugin_ptr is not initialised */
+				if (common_utility_plugin_ptr == NULL)
+					ereport(ERROR,
+							(errcode(ERRCODE_INTERNAL_ERROR),
+							errmsg("Failed to find common utility plugin.")));
+
+				special_func->valid_arg_types[i].valid_types_oid[j] = (*common_utility_plugin_ptr->get_tsql_datatype_oid)(special_func->valid_arg_types[i].valid_types[j]);
+			}
+
+			validTypId = special_func->valid_arg_types[i].valid_types_oid[j];
+
+			if (inputTypId == validTypId)
+			{
+				type_match = true;
+				break;
+			}
+		}
+		if (!type_match)
+		{
+			ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_FUNCTION),
+				 errmsg("Argument data type %s is invalid for argument %d of %s function.", 
+				 		 format_type_be(inputTypId), i+1, special_func->funcname)));
```

Comment:
Have we verified that we are throwing the error in Correct `Error Context`?

Reply:
When investigated about the `Error Context`  error context stack is not having correct value due to which crash happens when [error context callback is called](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/86f9c45d263a1da7ad16a7b60620452016410743/src/backend/utils/error/elog.c#L514) when reporting above error, this is due to a separate issue reported [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/issues/2693). This issue should be fixed under https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/397